modules/reactor/src/meshgenerators/AdvancedConcentricCircleGenerator.C

```diff
@@ -0,0 +1,487 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdvancedConcentricCircleGenerator.h"
+
+// C++ includes
+#include <cmath>
+
+registerMooseObject("ReactorApp", AdvancedConcentricCircleGenerator);
+
+InputParameters
+AdvancedConcentricCircleGenerator::validParams()
+{
+  InputParameters params = PolygonMeshGeneratorBase::validParams();
+  params.addRangeCheckedParam<unsigned int>(
+      "num_sectors",
+      "num_sectors>2",
+      "Number of azimuthal sectors of the circular mesh to be generated.");
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "customized_azimuthal_angles",
+      "customized_azimuthal_angles>=0&customized_azimuthal_angles<360",
+      "List of the user-specified azimuthal angles of the nodes.");
+  params.addRequiredRangeCheckedParam<std::vector<Real>>(
+      "ring_radii", "ring_radii>0", "Radii of major concentric circles (rings).");
+  params.addRequiredRangeCheckedParam<std::vector<unsigned int>>(
+      "ring_intervals",
+      "ring_intervals>0",
+      "Number of radial mesh intervals within each major concentric circle excluding their "
+      "boundary "
+      "layers.");
+  params.addRangeCheckedParam<std::vector<Real>>(
+      "ring_radial_biases",
+      "ring_radial_biases>0",
+      "Values used to create biasing in radial meshing for ring regions.");
```

Comment:
Not sure what is the best approach here. The `ring_radii` is required for ACCG but optional for PCCMG. It might lead to some complexity if we put them into a base class.

Reply:
Okay, will do. So, I should initialize the corresponding class variables in the derived classes, right?

Reply:
not if it can be done in the base class